server/tests/security/test_terminal_exec_tool.py

```
@@ -0,0 +1,410 @@
+    / "chat" / "backend" / "agent" / "tools" / "terminal_exec_tool.py"
+)
+
+
+def _load_function(name: str, extra_globals: dict | None = None):
+    """Extract a top-level function from the source file by name."""
+    tree = ast.parse(_SOURCE_FILE.read_text())
+    for node in tree.body:
+        if isinstance(node, ast.FunctionDef) and node.name == name:
+            module = ast.Module(body=[node], type_ignores=[])
+            namespace: dict = dict(extra_globals or {})
+            exec(compile(module, str(_SOURCE_FILE), "exec"), namespace)
```

Ruff (0.15.12)

[error] Use of `exec` detected

(S102)